Rebuild and allow user SSH

FrostyX opened the pull request:
Implementation of https://github.com/fedora-copr/debate/pull/6

Fixes: #2364

praiskup commented:
Per @lachmanfrantisek claim, @nforro wants to be the early adopter! :-)

FrostyX commented:
There is still work to be done but can you please do a preliminary review @praiskup? I don't want to sink time in fine-tuning the details without knowing, that the general idea looks good.

What remains to be done:

- There is a lot of valid linter errors that need to be fixed
- Obtaining SSH keys from FAS (or fallback to a general textarea) is not there yet
- I only tested in the docker-compose so far. Still requires some STG testing
- Small bugs (mainly `copr-builder show`)
- Some conceptual questions code TODOs, I am very much interested in your opinion

praiskup commented on backend/copr_backend/background_worker_build.py:
```diff
@@ -766,6 +857,9 @@ def build(self, attempt):
                              .format(transfer_failure))
+
+        self._keep_alive_for_user_ssh()
+        # TODO Should the results be downloaded for jobs with user SSH?
```
If we have a reasonable file-list filter, yes.  Otherwise, NO is safer option at the beginning.

praiskup commented on frontend/coprs_frontend/coprs/logic/builds_logic.py:
```diff
@@ -862,6 +865,8 @@ def add(cls, user, pkgs, copr, source_type=None, source_json=None,
             bootstrap=bootstrap,
             isolation=isolation,
+            allow_user_ssh=allow_user_ssh,
+            ssh_public_key=ssh_public_key,
```
Would it be possible to have this in YAML/JSON format or something...?   We
could have things like:
:
```yaml
---
type: FAS
fas_user: xyzuser
```

```
---
type: plaintext
pub_key: ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAACAQDj5HauAOur/6+...
```

```
---
type: online
url: https://github.com/praiskup.keys
```

We don't have to implement all the options in the web-UI right now, nor in cli.
But we can be "prepared".

FrostyX replied:
This sounds reasonable, not done yet.

FrostyX replied:
I thought about this and it is probably not needed now. We IMHO still want one big textarea for user input and then parse the different (types of) keys somewhere in our code. Personally, I would do the parsing on backend (we have workers there) or even on builder. Because we can fetch the keys from FAS or URL right after that.

Right now only plaintext and one key is supported, so the value could only be

```
ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAACAQDj5HauAOur/6+.
```

Seeing some other comments in the PR, we will probably implement support for multiple keys in this PR, so newline would be separator, e.g.

```
ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAACAQDj5HauAOur/6+.
ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAACAQDj5HauAOur/6+.
```

Once we start to support other methods, the input could be

```
ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAACAQDj5HauAOur/6+.
FAS:xyzuser
https://github.com/praiskup.keys
GitHub:frostyx
```

We can always save the blob to the database as is and parse it when needed. That would be consistent with how e.g. `copr://` external repositories work.

praiskup replied:
> Right now only plaintext and one key is supported

Huh, I've successfully tested before with multiple keys :-)

praiskup replied:
> We IMHO still want one big textarea for user input and then parse the different (types of) keys somewhere in our code.

I'd prefer to have a separate text area in the future, rather than changing the field and start "guessing" the type of input.

My thought was to start doing it right away in the background -- aka giving the user a simple textarea for multi-line key-only input, converting it to the yaml thing in the background, and store into DB.  This way we would have open doors for later enhancements without touching the old entries.  But I agree that this is rather a nice to have thing, there are many such things.

praiskup commented on frontend/coprs_frontend/coprs/models.py:
```diff
@@ -1444,6 +1450,8 @@ def sandbox(self):
             # build separated from any other.
             submitter = uuid.uuid4()
 
+        # TODO If allow_user_ssh, then add unique suffix
```
Yes please :)

FrostyX replied:
Done

praiskup commented on rpmbuild/bin/copr-builder:
```diff
@@ -0,0 +1,222 @@
+    like `if $(ssh stroj cat foo|head -n1) == "foo"`. So instead, we instruct
+    users to run `copr-builder help` manually.
+    """
+    # TODO Should this be in the documentation instead?
```
Not sure if "instead", but maybe "also"? :-)

FrostyX replied:
I mentioned the `copr-builder help` command in the Copr web UI, let's start from there. Maybe we will move it to the documentation page in the future.

praiskup commented on rpmbuild/bin/copr-builder:
```diff
@@ -0,0 +1,222 @@
+        "Please be responsible.\n"
+        "\n"
+        "This instance should be used only for the purposes of debugging\n"
+        "build failures of RPM packages. It is not a personal VPS :-).\n"
```
I think we could be more "strict", without a joking tone, ala motd on Batcave:
```
This is a PRIVATE computer system, unauthorized access is
strictly prohibited.

This system is to be used for Copr and related purposes only.  This
is not your personal computing system.  Users who are unsure whether or not
they have access to this system, don't have it and should log off immediately.
```

FrostyX replied:
Updated. I tried some middle ground between my half-joking tone and your super strict one :-)
I don't want to be too strict and provoke people to do bad stuff just to try if they can get away with it.

praiskup commented on rpmbuild/bin/copr-builder:
```diff
@@ -0,0 +1,222 @@
+from copr_common.helpers import timedelta_to_dhms
+
+
+RPMBUILD_PID_PATH = "/var/lib/copr-rpmbuild/pid"
```
We seem to have this in config already (helpers.py)

FrostyX replied:
Good catch, fixed.

praiskup commented on backend/copr_backend/background_worker_build.py:
```diff
@@ -740,6 +749,87 @@ def _add_pubkey(self):
+        """
+        Wait until user releases the VM or until it expires.
+        """
+        # TODO Should we use CancellableThreadTask for the implementation?
```
That's definitely nice to have.

praiskup commented on backend/copr_backend/background_worker_build.py:
```diff
@@ -740,6 +749,87 @@ def _add_pubkey(self):
+    def _deploy_user_ssh(self):
+        """
+        Deploy user public key to the builder, so that they can connect via SSH.
+        """
```
This is a nice first step, if we make sure there's no shell injection (shlex.quote again).  Long term it would be better to have a more controlled environment for root commands.

For some time we are facing subscription-manager headaches (subscribing also requires root access, and takes too much time during builder startup). It would be nice to be able to subscribe "on demand".  Or wait for the async subscription-manager action "on demand".  It would be nice to have a new "root" hook in resalloc doing these per-task preparation actions so we could drop the root access from Backend entirely.

FrostyX commented on frontend/coprs_frontend/coprs/auth.py:
```diff
@@ -139,6 +139,20 @@ def login():
         team_req = TeamsRequest(["_FAS_ALL_GROUPS_"])
+
+        # TODO Ask for SSH public key. Anybody knows where is the list of
+        # available keys that we can ask for?
```
I agree, I will create a separate RFE for this.

FrostyX reviewed (commented):
Thank you very much for the review. I updated almost everything but the PR is still WIP.

praiskup commented:
Not on a proper review yet, but a few notes after the first "hands on" [experience](https://copr.stg.fedoraproject.org/coprs/g/copr/TEST1701420778014423815-build-attributes/build/2916300/), some comments:
- I first wasn't sure where to ssh, I assume presenting the ssh command in the web-UI isn't a trivial (frontend has no idea about the machine IP/host), but dumping somewhere "check backend.log for more instructions" would is IMO necessary
- the build ended up in the `failed` state (OK) but the source build is still "running", and the machine has already been dealocated (I did not prolong it)
- the log [disappeared](https://download.copr-dev.fedorainfracloud.org/results/@copr/TEST1701420778014423815-build-attributes/srpm-builds/02916300/backend.log.gz) - which is again expected, but we have a non-working link there

praiskup commented:
- no motd ...
- backend.log contains no info about "prolonging"

praiskup commented:
- I noticed the instructions say "Public Key" but we should say "Public Keys" - making sure that users know that multiple keys may be passed

praiskup commented on backend/copr_backend/helpers.py:
```diff
@@ -240,6 +240,8 @@ def _get_limits_conf(parser):
     limits['owner'] = _get_conf(
         parser, "backend", "builds_max_workers_owner", 20, mode="int")
+    limits['userssh'] = _get_conf(
+        parser, "backend", "builds_max_userssh", 2, mode="int")
```
Might be tricky; we should warn the user that, when the ssh key is specified, only 2 parallel builders will be given to him and that the rest of builds stays in "pending" state.

praiskup commented on backend/copr_backend/rpm_builds.py:
```diff
@@ -150,6 +151,25 @@ def __init__(self, architecture, limit):
+            # We don't allow user SSH for SRPM builds, return unique string
+            # making this unlimited
+            if x.source_build:
+                return str(uuid.uuid4())
```
I think we can simply return None, and the HashWorkerLimit will not count with those.
Btw., I noticed this because of the memory leak we have in StringCounter() (at least with random strings like this).

praiskup commented on backend/copr_backend/rpm_builds.py:
```diff
@@ -150,6 +151,25 @@ def __init__(self, architecture, limit):
+                return str(uuid.uuid4())
+
+            # Not limit builds that doesn't allow user SSH
+            # pylint: disable=protected-access
```
nit: "Don't limit builds that don't allow users to SSH"

praiskup commented on backend/copr_backend/sshcmd.py:
```diff
@@ -243,8 +244,23 @@ def _rsync_download(self, src, dest, logfile, subprocess_timeout):
+        for value in include or []:
+            command.extend(["--include", shlex.quote(value)])
+        for value in exclude or []:
+            command.extend(["--exclude", shlex.quote(value)])
```
Hm, I'm not sure I like the `include` and `exclude` split in this method.  The semantics of these options in rsync is order dependant;  it would be nicer if we had just one argument named `filter`  to have more flexibility for the future.

praiskup commented on backend/tests/test_config_reader.py:
```diff
@@ -65,6 +66,7 @@ def test_correct_build_limits(self):
                 'ppc64le': 11,
                 's390x': 5,
             },
+            'userssh': 2,
```
Can you add a config above, and change this to a non-default value?

praiskup commented on frontend/coprs_frontend/alembic/versions/41763f7a5185_add_allow_user_ssh_column.py:
```diff
@@ -0,0 +1,27 @@
+def upgrade():
+    op.add_column('build', sa.Column('allow_user_ssh', sa.Boolean(),
+                                     server_default='0', nullable=False))
+    op.add_column('build', sa.Column('ssh_public_key', sa.Text()))
```
Do we need both fields?

praiskup commented on frontend/coprs_frontend/coprs/models.py:
```diff
@@ -1440,13 +1445,20 @@ def sandbox(self):
-        return '{0}--{1}'.format(self.copr.full_name, submitter)
+        # If we don't know build submitter, use "random" value and keep the
+        # build separated from any other.
+        submitter = self.submitter[0] or str(uuid.uuid4())
```
is submitter[0] correct? not submitter?

praiskup replied:
I was wrong, submitter[0] is correct.

praiskup commented on frontend/coprs_frontend/coprs/models.py:
```diff
@@ -1440,13 +1445,20 @@ def sandbox(self):
+            # an unique sandbox
+            is_rpm_build = self.source_state in helpers.FINISHED_STATES
+            if self.allow_user_ssh and is_rpm_build:
+                parts.append(str(uuid.uuid4()))
```
This is imo unnecessarily complicated, we could somewhere at the beginning do:
```
if self.allow_user_ssh:
    submitter = str(uuid.uuid4())
```

praiskup commented on frontend/coprs_frontend/coprs/templates/coprs/detail/_builds_forms.html:
```diff
@@ -215,8 +215,21 @@ <h3 class="panel-title">
+      <code>copr-builder help</code> for complete instructions.
+    </p>
+    <input type="hidden" name="allow_user_ssh" value="{{ allow_user_ssh }}">
+    {{ render_field(form.ssh_public_key, label='Public SSH key', rows=8, cols=50) }}
```
s/key/keys/, (newline separated)

FrostyX commented:
Thank you very much for the review @praiskup,
I addressed all your comments. Everything is in a separate commit for the reviewing convenience, I will squash them before merging.

> no motd ...

I spent a couple of hours on this before Christmas and wasn't able to set up MOTD on the builders. And then @xsuchy suggested it might be a good idea not having MOTD because of cases like `if $(ssh ... cat foo|head -n1) == "foo"`. So I put in the instructions to run `copr-builder help` instead.

praiskup commented:
> if $(ssh ... cat foo|head -n1) == "foo"

But MOTD shouldn't affect stdout:

```
$ ssh batcave01.iad2.fedoraproject.org echo ahoj
ahoj
```

praiskup commented:
jakub to ping pavel and take a look at motd problems..

praiskup commented on backend/copr_backend/background_worker_build.py:
```diff
@@ -307,6 +318,9 @@ def _parse_results(self):
         """
         Parse `results.json` and update the `self.job` object.
         """
+        if self.job.ssh_public_keys:
```
The implication here deserves an inline comment :-)

praiskup commented on backend/copr_backend/background_worker_build.py:
```diff
@@ -683,6 +704,9 @@ def _collect_built_packages(self, job):
         self.log.info("Listing built binary packages in %s", job.results_dir)
 
+        if self.job.ssh_public_keys:
+            return "Public SSH access was enabled, no results were stored"
```
Isn't it safer to return empty dict here?

praiskup commented on backend/copr_backend/sshcmd.py:
```diff
@@ -243,8 +243,21 @@ def _rsync_download(self, src, dest, logfile, subprocess_timeout):
+            "{}/".format(dest),
+            "&>", log_filepath,
+        ])
+        command = " ".join(command)
```
Nice!

praiskup commented on backend/copr_backend/rpm_builds.py:
```diff
@@ -150,6 +150,25 @@ def __init__(self, architecture, limit):
+    """
+    Limit the number of builders that allow user SSH
+    """
+    def __init__(self, _hasher, limit):
```
You can drop the unused hasher argument I think.

praiskup commented on frontend/coprs_frontend/coprs/templates/coprs/detail/_builds_forms.html:
```diff
@@ -215,8 +215,31 @@ <h3 class="panel-title">
+    <input type="hidden" name="allow_user_ssh" value="{{ allow_user_ssh }}">
+    {{ render_field(form.ssh_public_keys, label='Public SSH keys',
+                    placeholder='Newline separated public SSH keys, e.g. ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQDR+QU9...',
+                    rows=8, cols=50) }}
```
This seems damn awesome! :-) Can't wait for testing it.

praiskup commented:
Small style issue:
![Screenshot_20240118_164643](https://github.com/fedora-copr/copr/assets/3277839/dd4b3f08-246f-42cc-a862-982857cdf56b)

praiskup commented:
I'm a bit uncertain about the windows-like EOLs, in vim I see:
```
... praiskup.usersys.redhat.com^M
```

praiskup commented:
I tried to drop `.ssh/authorized_keys` file - checking what happens :-) - and it just finished the otherwise succeeding build.  Is this OK?

praiskup commented on backend/copr_backend/background_worker_build.py:
```diff
@@ -51,6 +58,9 @@
     "rpm_q_builder": "rpm -q copr-rpmbuild --qf \"%{VERSION}\n\"",
+    "echo_authorized_keys": "echo {0} >> /root/.ssh/authorized_keys",
+    "set_expiration": "echo -n {0} > " + USER_SSH_EXPIRATION_PATH,
+    "cat_expiration": "cat {0}".format(USER_SSH_EXPIRATION_PATH),
```
I think we should document here that we must be especially careful about error handling; users may replace the `echo` or `cat` commands, and (in theory) cause damage to the backend machine. This deserves documenting IMO (here ore in a later PR).

praiskup commented on backend/copr_backend/background_worker_build.py:
```diff
@@ -589,12 +605,19 @@ def _download_results(self):
+        if self.job.ssh_public_keys:
+            self.log.info("Builder allowed user SSH, not downloading the "
+                          "results for safety reasons.")
+            filter_ = ["+ success", "+ *.spec", "- *"]
```
I'm thinking ... what if user replaces the remote side (server) `rsync` binary.  Can that trick the backend to download some unwanted stuff?

Ideas?

FrostyX replied:
This looks quite dangerous. Should we maybe verify some checksum of the binary, or something like that? Same for the `cat` commands mentioned above? But maybe that leads to the same issue (i.e. we can't be sure that the checksum tool wasn't tampered with)?

praiskup commented:
Just a few comments above, otherwise it seems OK to squash and merge. The motd and other things can be done later.

FrostyX commented:
> Small style issue:

Do you want the instructions in the "SSH access to the builder" section to be indented to the right?

> I'm a bit uncertain about the windows-like EOLs, in vim I see:
> ... praiskup.usersys.redhat.com^M

Hmm, I haven't encountered this. Where do you see the ^M characters?

FrostyX commented on backend/copr_backend/background_worker_build.py:
```diff
@@ -740,6 +766,122 @@ def _add_pubkey(self):
+        """
+        Find the user preference for the builder expiration.
+        """
+        rc, out, _err = self.root_ssh.run_expensive(COMMANDS["cat_expiration"])
```
Don't use `run_expensive` and check output length and/or timeout.

FrostyX replied:
I couldn't use `run` because it doesn't return the output. So I kept `run_expensive` and added a timeout for it.

FrostyX commented:
> Small style issue:
![Screenshot_2024-01-26_12-55-04](https://github.com/fedora-copr/copr/assets/2151771/76204149-d779-4f51-8ed5-014920dabdc9)

I changed the style a bit.

praiskup reviewed (approved):
Thank you!
